### ISSUE TITLE
Godot 4.0 Compailing error vulkan\\loader unknown type name 'IDXGIFactory6'

### ISSUE THREAD

**Issue by Al-Shahrior** [godotengine/godot#45100]
**Godot version:**

Godot #4.0
commit 2830a152de98bb8e1f09a3228375028699f033ef

**OS/device including version:**

Windows 10 19009 / Intel graphics

**Issue description:**

Trying to compile the master branch on windows 10. Using gcc 9. I'm new on Godot. I do Apologize if make any funny action. I'm a bit excited about the Godot 4.0
![image](https://user-images.githubusercontent.com/31450209/104186083-144a8400-5440-11eb-89a9-3f51e57832d3.png)
```
=====
b"thirdparty\\vulkan\\loader\\loader.c: In function 'loader_create_instance_chain':\nthirdparty\\vulkan\\loader\\loader.c:6059:5: error: unknown type name 'IDXGIFactory6'; did you mean 'IDXGIFactory1'?\n 6059 |     IDXGIFactory6* dxgi_factory = NULL;\n      |     ^~~~~~~~~~~~~\n      |     IDXGIFactory1\nthirdparty\\vulkan\\loader\\loader.c:6060:42: error: 'IID_IDXGIFactory6' undeclared (first use in this function); did you mean 'IID_IDXGIFactory5'?\n 6060 |     HRESULT hres = fpCreateDXGIFactory1(&IID_IDXGIFactory6, (void **)&dxgi_factory);\n      |                                          ^~~~~~~~~~~~~~~~~\n      |                                          IID_IDXGIFactory5\nthirdparty\\vulkan\\loader\\loader.c:6060:42: note: each undeclared identifier is reported only once for each function it appears in\nthirdparty\\vulkan\\loader\\loader.c:6063:21: error: request for member 'lpVtbl' in something not a structure or union\n 6063 |         dxgi_factory->lpVtbl->Release(dxgi_factory);\n      |                     ^~\nthirdparty\\vulkan\\loader\\loader.c: In function 'ReadSortedPhysicalDevices':\nthirdparty\\vulkan\\loader\\loader.c:7088:5: error: unknown type name 'IDXGIFactory6'; did you mean 'IDXGIFactory1'?\n 7088 |     IDXGIFactory6* dxgi_factory = NULL;\n      |     ^~~~~~~~~~~~~\n      |     IDXGIFactory1\nthirdparty\\vulkan\\loader\\loader.c:7089:42: error: 'IID_IDXGIFactory6' undeclared (first use in this function); did you mean 'IID_IDXGIFactory5'?\n 7089 |     HRESULT hres = fpCreateDXGIFactory1(&IID_IDXGIFactory6, (void **)&dxgi_factory);\n      |                                          ^~~~~~~~~~~~~~~~~\n      |                                          IID_IDXGIFactory5\nthirdparty\\vulkan\\loader\\loader.c:7106:32: error: request for member 'lpVtbl' in something not a structure or union\n 7106 |             hres = dxgi_factory->lpVtbl->EnumAdapterByGpuPreference(dxgi_factory, i, DXGI_GPU_PREFERENCE_UNSPECIFIED, &IID_IDXGIAdapter1, (void **)&adapter);\n      |                                ^~\nthirdparty\\vulkan\\loader\\loader.c:7106:86: error: 'DXGI_GPU_PREFERENCE_UNSPECIFIED' undeclared (first use in this function)\n 7106 |             hres = dxgi_factory->lpVtbl->EnumAdapterByGpuPreference(dxgi_factory, i, DXGI_GPU_PREFERENCE_UNSPECIFIED, &IID_IDXGIAdapter1, (void **)&adapter);\n      |                                                                                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\nthirdparty\\vulkan\\loader\\loader.c:7190:21: error: request for member 'lpVtbl' in something not a structure or union\n 7190 |         dxgi_factory->lpVtbl->Release(dxgi_factory);\n      |                     ^~\n"
=====
scons: *** [thirdparty\vulkan\loader\loader.windows.tools.64.o] Error 1
scons: building terminated because of errors.
```

**Steps to reproduce:**

**Minimal reproduction project:**

**Comment by Calinou** [comment]
Duplicate of https://github.com/godotengine/godot/issues/44637

**Comment by Al-Shahrior** [comment]
I'm using gcc (tdm64-1) 9.2.0
![image](https://user-images.githubusercontent.com/31450209/104188784-fe3ec280-5443-11eb-92d0-e348f1780ad9.png)
Then why i'm still getting error ?

**Comment by Al-Shahrior** [comment]
@Calinou I'm using gcc (tdm64-1) 9.2.0

**Comment by Xartorx** [comment]
Try [this](http://winlibs.com/) version of GCC, latest one.

**Comment by Calinou** [comment]
@Al-Shahrior I'm referring to the MinGW version, not the GCC version. GCC 9.x on Windows is paired with MinGW 7.0, which is too old to compile the `master` branch.

**Comment by Al-Shahrior** [comment]
Привет @Xartorx  Thanks a lot. I managed to compile the master branch. It's took about 2.1 hour. And 1.09GB. Without WebM SIMD optimizations.

**Comment by Calinou** [comment]
> I managed to compile the master branch. It's took about 2.1 hour. And 1.09GB. Without WebM SIMD optimizations.

Linking with MinGW on Windows is quite slow. For development purposes, I'd recommend using MSVC instead.

**Comment by bruvzg** [comment]
> I managed to compile the master branch. It's took about 2.1 hour. And 1.09GB.

Or LLVM+MinGW tool chain, e.g. https://github.com/mstorsjo/llvm-mingw, LLD is much faster than GNU ld.

> Without WebM SIMD optimizations.

To enable optimizations, add https://yasm.tortall.net/ executable to your PATH.

**Comment by Xartorx** [comment]
> Or LLVM+MinGW tool chain, e.g. https://github.com/mstorsjo/llvm-mingw, LLD is much faster than GNU ld.


Didn't worked for me - went on 25+ minute linking process until LLD ran out of memory and crashed.
And with mono build it crashed with error.
<details>
    <summary>Error:</summary>

>[ 81%] b'lld-link: error: undefined symbol: thread-local initialization routine for GDMonoUtils::current_invoke_count
referenced by modules\\mono\\csharp_script.cpp
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;libmodule_mono.windows.opt.tools.64.a(csharp_script.windows.opt.tools.64.o)
referenced by modules\\mono\\mono_gd\\gd_mono_marshal.cpp
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;libmodule_mono.windows.opt.tools.64.a(gd_mono_marshal.windows.opt.tools.64.o)
referenced by modules\\mono\\signal_awaiter_utils.cpp
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;libmodule_mono.windows.opt.tools.64.a(signal_awaiter_utils.windows.opt.tools.64.o)
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
'
=====

</details>

Yeah, safest bet is using MSVC on Windows.

**Comment by slapin** [comment]
@Al-Shahrior which version ow mingw worked for you?

**Comment by Al-Shahrior** [comment]
@slapin This version seem working.
This is the winlibs 64-bit standalone build of:
- GCC 10.2.0
- GDB 10.1
- LLVM/Clang/LLD/LLDB 11.0.0
- MinGW-w64 8.0.0
- GNU Binutils 2.35.1
- GNU Make 4.3
- PExports 0.47
- dos2unix 7.4.2
- Yasm 1.3.0
- NASM 2.15.05
- JWasm 2.12pre

### But whenever i use this `use_lto=yes` parameter i'm getting linking error don't no  if it's compiler issue.
 I have open a new issue  for that #45470
